certbot-apache/certbot_apache/interfaces.py

```diff
@@ -123,10 +123,19 @@ class ParserNode(object):
     dirty: bool
 
     # Filepath of the file where the configuration element for this ParserNode
-    # object resides.
+    # object resides. This can be None if a configuration directive is defined in
+    # for example the httpd command line.
     filepath: Optional[str]
     """
 
+    @abc.abstractmethod
+    def __init__(self, ancestor=None, filepath="", dirty=False):  # pragma: no cover
```

Comment:
I have my same feeling as I did at https://github.com/certbot/certbot/pull/7308#discussion_r314111032 where I'm somewhat tempted to not give `ancestor` a default since not having one is uncommon, but you can ignore that if you prefer.

I feel the same way about `filepath`, but if you prefer to give it one, I think it should be `None` and not the empty string.

These comments apply to some of the derived classes as well.

Reply:
I agree and made both of these required now.